### PR TITLE
drivers: sensor: max30101: Augment driver to be trigger based

### PR DESCRIPTION
**Pull request by trembel**
This pull request contains different fixes and additions:
- Change from the old driver structure to the `dt_spec` structure
- Change from Kconfig based structure to using the devicetree, in order to support multiple instances of the same sensor
- Augmentation to support interrupts of the MAX30101
- Fixes on not forwarding return messages of i2c transactions

Signed-off-by: Silvano Cortesi <silvano.cortesi@hotmail.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
@MaureenHelm @gmarull @avisconti could you review?

File: drivers/sensor/max30101/max30101.h [hunk 1]
```diff
@@ -49,40 +64,48 @@
...
 
 enum max30101_pw {
-	MAX30101_PW_15BITS		= 0,
-	MAX30101_PW_16BITS,
-	MAX30101_PW_17BITS,
-	MAX30101_PW_18BITS,
+        MAX30101_PW_15BITS		= 0,
+        MAX30101_PW_16BITS,
+        MAX30101_PW_17BITS,
+        MAX30101_PW_18BITS,
```
Comment (gmarull) [inline drivers/sensor/max30101/max30101.h:108]: why these changes?

File: drivers/sensor/max30101/max30101_trigger.c [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+	if (ret < 0) {
+		LOG_ERR("Failed to set gpio as input!");
+		return ret;
+	}
+
+	gpio_init_callback(&data->gpio_cb, max30101_gpio_callback, BIT(config->gpio_int.pin));
+
+	ret = gpio_add_callback(config->gpio_int.port, &data->gpio_cb);
+	if (ret < 0) {
```
Comment (gmarull) [inline drivers/sensor/max30101/max30101_trigger.c:210]: please be consistent with blank lines:

```c
	ret = gpio_add_callback(config->gpio_int.port, &data->gpio_cb);
	if (ret < 0) {
```

not needed here:
```c
	int ret = gpio_pin_configure_dt(&config->gpio_int, GPIO_INPUT);

	if (ret < 0) {
```

File: drivers/sensor/max30101/Kconfig [hunk 3]
```diff
@@ -9,195 +9,4 @@ menuconfig MAX30101
...
-	  2: LED2 (IR), LED2_PA
-	  3: LED3 (green), LED3_PA
-	  4: None (disabled)
-	  5: LED1 (red), PILOT_PA
-	  6: LED2 (IR), PILOT_PA
-	  7: LED3 (green), PILOT_PA
-
-endif # MAX30101_MULTI_LED_MODE
-
 endif # MAX30101
```
Comment (avisconti) [inline drivers/sensor/max30101/Kconfig:60]: So Kconfig is empty? Why not removing it?

File: drivers/sensor/max30101/max30101.c [hunk 4]
```diff
@@ -168,22 +168,22 @@ static int max30101_init(const struct device *dev)
...
-	}
-	if (i2c_reg_write_byte_dt(&config->bus,
-			       MAX30101_REG_MULTI_LED + 1, multi_led[1])) {
-		return -EIO;
+		if (i2c_reg_write_byte_dt(&config->bus,
+			      MAX30101_REG_MULTI_LED, multi_led[0])) {
+			return -EIO;
+		}
+		if (i2c_reg_write_byte_dt(&config->bus,
+			      MAX30101_REG_MULTI_LED + 1, multi_led[1])) {
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/max30101/max30101.c]: exactly

File: dts/bindings/sensor/maxim,max30101.yaml [hunk 5]
```diff
@@ -1,8 +1,250 @@
...
+      default: 2
+      enum:
+       - 0
+       - 1
+       - 2
+       - 3
+       - 4
+       - 5
+       - 6
+       - 7
```
Comment (avisconti) [inline dts/bindings/sensor/maxim,max30101.yaml:130]: Maybe you can add comments for each case here, instead of in the description.
BTW, why the description has only 4 cases?

File: dts/bindings/sensor/maxim,max30101.yaml [hunk 6]
```diff
@@ -1,8 +1,250 @@
...
+      default: 0
+      enum:
+       - 0
+       - 1
+       - 2
+       - 3
+       - 4
+       - 5
+       - 6
+       - 7
```
Comment (avisconti) [inline dts/bindings/sensor/maxim,max30101.yaml:150]: ditto. 
I think is more clear to add the enum values description in a comment for each line above.
Please do it for all.

File: dts/bindings/sensor/maxim,max30101.yaml [hunk 7]
```diff
@@ -1,8 +1,250 @@
...
+      description: |
+        Set which LED and pulse amplitude are active in time slot 1.
+        0 = None (disabled)
+        1 = LED1 (red), LED1_PA
+        2 = LED2 (IR), LED2_PA
+        3 = LED3 (green), LED3_PA
+        4 = None (disabled)
+        5 = LED1 (red), PILOT_PA
+        6 = LED2 (IR), PILOT_PA
+        7 = LED3 (green), PILOT_PA
```
Comment (avisconti) [inline dts/bindings/sensor/maxim,max30101.yaml:186]: here 7, but only 6 in the enum.

File: dts/bindings/sensor/maxim,max30101.yaml [hunk 8]
```diff
@@ -1,8 +1,250 @@
...
+      description: |
+        Set which LED and pulse amplitude are active in time slot 2.
+        0 = None (disabled)
+        1 = LED1 (red), LED1_PA
+        2 = LED2 (IR), LED2_PA
+        3 = LED3 (green), LED3_PA
+        4 = None (disabled)
+        5 = LED1 (red), PILOT_PA
+        6 = LED2 (IR), PILOT_PA
+        7 = LED3 (green), PILOT_PA
```
Comment (avisconti) [inline dts/bindings/sensor/maxim,max30101.yaml:209]: 7 but only 6 in the enum

File: dts/bindings/sensor/maxim,max30101.yaml [hunk 9]
```diff
@@ -1,8 +1,250 @@
...
+      description: |
+        Set which LED and pulse amplitude are active in time slot 3.
+        0 = None (disabled)
+        1 = LED1 (red), LED1_PA
+        2 = LED2 (IR), LED2_PA
+        3 = LED3 (green), LED3_PA
+        4 = None (disabled)
+        5 = LED1 (red), PILOT_PA
+        6 = LED2 (IR), PILOT_PA
+        7 = LED3 (green), PILOT_PA
```
Comment (avisconti) [inline dts/bindings/sensor/maxim,max30101.yaml:232]: ditto

File: dts/bindings/sensor/maxim,max30101.yaml [hunk 10]
```diff
@@ -1,8 +1,250 @@
...
+      description: |
+        Set which LED and pulse amplitude are active in time slot 4.
+        0 = None (disabled)
+        1 = LED1 (red), LED1_PA
+        2 = LED2 (IR), LED2_PA
+        3 = LED3 (green), LED3_PA
+        4 = None (disabled)
+        5 = LED1 (red), PILOT_PA
+        6 = LED2 (IR), PILOT_PA
+        7 = LED3 (green), PILOT_PA
```
Comment (avisconti) [inline dts/bindings/sensor/maxim,max30101.yaml:255]: ditto

If you put the comments in the enum cases as suggested, I think that it is a good way also to catch the mismatch error

File: samples/sensor/max30101/src/main.c [hunk 11]
```diff
@@ -24,10 +24,10 @@ void main(void)
...
-		sensor_channel_get(dev, SENSOR_CHAN_GREEN, &green);
+		sensor_channel_get(dev, SENSOR_CHAN_RED, &red);
 
-		/* Print green LED data*/
-		printf("GREEN=%d\n", green.val1);
+		/* Print red LED data*/
+		printf("RED=%d\n", red.val1);
 
 		k_sleep(K_MSEC(20));
 	}
```
Comment (avisconti) [inline samples/sensor/max30101/src/main.c:62]: I think this commit should be squashed with the driver one. This one alone (w/o the multi-instance driver) has no sense. The commit is an atomic piece of change that spans across multiple parts in the projects (driver, sample, dts bindings, test, ...)

**Review by avisconti — CHANGES_REQUESTED** [review]
